Rename FP_SHARING and its friends to FPU_SHARING

stephanosio opened the pull request:
This PR renames all `FP_SHARING` and the related names to `FPU_SHARING`, in order to remove any ambiguity as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/24636#issuecomment-618993135.

stephanosio committed:
kconfig: Rename CONFIG_FP_SHARING to CONFIG_FPU_SHARING

This commit renames the Kconfig `FP_SHARING` symbol to `FPU_SHARING`,
since this symbol specifically refers to the hardware FPU sharing
support by means of FPU context preservation, and the "FP" prefix is
not fully descriptive of that; leaving room for ambiguity.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kconfig: Rename x86 FPU sharing symbols

This commit renames the x86 Kconfig `CONFIG_{EAGER,LAZY}_FP_SHARING`
symbol to `CONFIG_{EAGER,LAZY}_FPU_SHARING`, in order to align with the
recent `CONFIG_FP_SHARING` to `CONFIG_FPU_SHARING` renaming.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: Rename arch.arm.swap.common.fp_sharing

This commit renames the `arch.arm.swap.common.fp_sharing` tests to
`arch.arm.swap.common.fpu_sharing`, in order to align with the recent
`CONFIG_FP_SHARING` to `CONFIG_FPU_SHARING` renaming.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: Rename kernel.common.stack_protection_arm_fp_sharing

This commit renames the `kernel.common.stack_protection_arm_fp_sharing`
test to `kernel.common.stack_protection_arm_fpu_sharing`, in order to
align with the recent `CONFIG_FP_SHARING` to `CONFIG_FPU_SHARING`
renaming.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: Rename kernel.fp_sharing

This commit renames the `kernel.fp_sharing` tests to
`kernel.fpu_sharing`, in order to align with the recent
`CONFIG_FP_SHARING` to `CONFIG_FPU_SHARING` renaming.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
cc @pfalcon @pabigot

ioannisg reviewed (commented):
This is probably fine, although the improvement is not as significant as when we renamed `config FLOAT` to `config FPU`: `config FP_SHARING` signifies that the FP Register list is shared between threads, so if we want to go for a fully descriptive symbol name, that should be `config FP_REG_SHARING` or something similar. Effectively, all the option does is to force thread switch to save and restore the content of the FP Registers. 
FPU_SHARING is not making things more clear.

ioannisg commented:
I think if we take this effort, we should rename the symbol to something that explicitly describes the functionality. But I won't block this, if other ARCH maintainers want to accept the patch as is.

pfalcon commented:
@ioannisg: In the comment which started it, which is also quoted in the description: https://github.com/zephyrproject-rtos/zephyr/pull/24636#issuecomment-618993135 , the talk was about: a) consistency, b) non-confusion. There never has been talk about perfectionism in naming/descriptions, only about non-messiness in naming/description. And I don't think that adding perfectionism or scholasticism a-la "it's not floating-point unit, it's floating point unit registers!" would help.

So, let's get back to "a) consistency, b) non-confusion" criteria and see how the rename "CONFIG_FP_SHARING -> CONFIG_FPU_SHARING" matches the desired direction:

1. Consistency: there was CONFIG_FPU which enabled FPU, and an option which modified aspects of that option behavior. It's natural to consider such an option as a "suboption" of the original option, and natural way to name such options is CONFIG_&lt;MAINOPTION>_&lt;SUBOPTION>_&lt;SUBSUBOPTION>_... . Hence, CONFIG_FPU_SHARING.
2. Non-confusion: "FP sharing" would mean "floating point sharing", and it's not exactly clear what it means. Whereas "FPU sharing" is "floating point unit sharing", and pretty clearly describes what happens there.

So, I appreciate @stephanosio's coming forward with this refactor, even though I have to admit I'm surprised by the amount of changes it took. Fortunately, with this done, we can feel that we did our (realistic) best for our users, and can call it done (IMHO).

pfalcon reviewed (approved):
Thanks for coming forward with this!

ioannisg commented:
@pfalcon 
Regarding (1): that's fair point.
Regarding (2): fair as well; btw, I never proposed to keep the original FP_SHARING name, so that's a bit out-dated :)
How about `config FPU_REGS_SHARING`, which satisfies (1) and (2) and makes it 100% clear what part of FPU is exactly shared? Would help a to understand the FP-regs save-restore in thread swap code

stephanosio commented:
After all, whether it is the FPU registers or some other form of FPU-related context, we are sharing one or more hardware floating point unit(s) among multiple threads, and IMHO, the `FPU_SHARING` nomenclature is clear enough in that aspect (at least, we are now clearly referring to the "floating point unit" instead of the rather abstract entity called "floating point").

Also it is worth noting that we are preserving per-thread "floating point unit context", which happens to be, but not necessarily always, the FPU registers in most architectures.

pfalcon commented:
@ioannisg:

> How about config FPU_REGS_SHARING, which satisfies (1) and (2) and makes it 100% clear what part of FPU is exactly shared?

My biggest concern is saving @stephanosio from frustration. The situation as it is now is: I (personally) talked @stephanosio into going for this refactor, which turned out to be changing 67 files around. Now that it was done, we're going into bikeshedding mode, discussing whether it *could* have been changed into something else. I'd find such direction to be frustrating. The choice for "FPU_SHARING" wasn't random, as I explained. And IMHO, it's not worth to spend more time/effort trying to get diminishing returns of making it "even better by 2%".

Even in pure context-free manner, looking at CONFIG_FPU_REGS_SHARING vs CONFIG_FPU_SHARING, I see that 1st is long, as in, "too long" (and that puts quite realistic limits of how descriptive we can go; again, the aim is not to be "as descriptive as possible", it's "being descriptive enough, promote consistency, and avoid confusion").

And I agree with @stephanosio comment above that adding "REGS" doesn't add much info, after all, FPU registers are part of FPU overall.

stephanosio commented:
@andrewboie @andyross @dcpleung @vonhust @carlocaione @nategraff-sifive Some review/feedback from the arch people would be greatly appreciated, as, if we are ever going to get this in, it should be for 2.3.

tyrantlizard77 reviewed (approved):
> @andrewboie @andyross @dcpleung @vonhust @carlocaione @nategraff-sifive Some review/feedback from the arch people would be greatly appreciated, as, if we are ever going to get this in, it should be for 2.3.

I'll be honest: I don't particularly care what it's called.
The discussion in this PR feels like the textbook definition of a bikeshed.
Merging in its current form has no objections from me.

ioannisg commented:
On a side note, this is a "user" option (part of the public API), likely used in downstream Zephyr applications; do we have a policy for renaming these?
@carlescufi ^^

ruuddw reviewed (approved):
FPU_SHARING is fine.

pfalcon commented:
@ioannisg 
> On a side note, this is a "user" option (part of the public API), likely used in downstream Zephyr applications; do we have a policy for renaming these?

Just the same as the previous rename - https://github.com/zephyrproject-rtos/zephyr/pull/24636. I had to go thru Zephyr ports of projects we have in CI, and update them. And I'm surprised that no concerns about backwards compatibility were raised during review of that PR. I normally would complain about breaking changes, but decided to save on that, as overall intention was "good".

And finally, as my comment in the description of this PR explains, I exactly come up with suggestion to rename FP_SHARING -> FPU_SHARING to batch such kind of changes and limit them to a single Zephyr release, and hopefully not come up to it anytime soon.

ioannisg commented:
Agree with you @pfalcon that since we broke compatibility already, better to rename this option in this release cycle. I just wanted to raise awareness here, that we cannot be simply renaming public Kconfig "APIs"; we need to establish a policy (doesn't look we have one at the moment; we could alternatively extend existing API-change policies to include Kconfig)

tyrantlizard77 reviewed (dismissed):
I've changed my mind. This is a pedantic change which is going to break people's Kconfigs. No thanks.

stephanosio commented:
> I've changed my mind. This is a pedantic change which is going to break people's Kconfigs. No thanks.

@andrewboie The main concern with this change would be the off-tree boards that set `CONFIG_FP_SHARING=y`.

Since `CONFIG_FP_SHARING=y` must be set alongside `CONFIG_FLOAT=y` and we have renamed `CONFIG_FLOAT` to `CONFIG_FPU` (which, by the way, had a solid enough rationale) for 2.3, I believe the extra effort by users to update `CONFIG_FLOAT` alongside `CONFIG_FP_SHARING` should be minimal (note that almost all board conf files do `CONFIG_FP_SHARING=y` right after `CONFIG_FLOAT=y`) -- which is also why we are trying to introduce this arguably pedantic change in none other than this release.

andrewboie commented:
> Since `CONFIG_FP_SHARING=y` must be set alongside `CONFIG_FLOAT=y` and we have renamed `CONFIG_FLOAT` to `CONFIG_FPU` (which, by the way, had a solid enough rationale) for 2.3,

A "this ship has already sailed" argument I suppose is good enough. I didn't realize we already broke end users with a prior change. I still think this is all a waste of time but I'll remove my -1.

stephanosio commented:
> I still think this is all a waste of time

@andrewboie As for renaming `FP_SHARING` to `FPU_SHARING`, for the most part, I agree.

Renaming `CONFIG_FLOAT` to `CONFIG_FPU`, I believe, was worth the effort; at least, it would potentially save a substantial amount of wasted time and effort arising from misunderstanding/confusion for many in the future (as I have already done in #23948).

With that, `FP_SHARING` was pretty much up for grabs, so here we are :)

ioannisg commented:
> A "this ship has already sailed" argument I suppose is good enough.

That's what we've discussed above in the comments timeline. I guess we can merge this PR since there's no additional harm it can do. From now on I think we really need to communicate public Kconfig changes to the API working group that seems to handling APIs in Zephyr.

FYI @carlescufi

stephanosio commented:
> From now on I think we really need to communicate public Kconfig changes to the API working group that seems to handling APIs in Zephyr.

For the starters, it would be nice to document these changes under the current "API Changes" section (but is this really an "API"?) or a new "Configuration Changes (?)" section in the release notes.

Also we should introduce some kind of deprecation policy/mechanism (e.g. warning), so that we do not immediately break things for users in a new release. I wonder if there is a good way to handle that in Kconfig (cc @ulfalizer).

mbolivar-nordic commented:
> The problem is the downstream code where this check is not available. Maybe we should consider making `KconfigCheck` (and potentially others) a west extension so users can run it on their forks and out of tree code?

How is this unavailable to downstreams? The ci-tools repository is part of west.yml.

stephanosio commented:
> How is this unavailable to downstreams? The ci-tools repository is part of west.yml.

I meant "not easily accessible" (unless it is already available as a west extension or possibly CMake target?). I doubt many users are aware of the existence of `check_compliance.py` and/or the fact that you can run it on a local repo. Also, AFAICS, the `check_compliance.py` does not support running on out-of-tree code, does it?

mbolivar-nordic commented:
> Also, AFAICS, the `check_compliance.py` does not support running on out-of-tree code, does it?

https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/Jenkinsfile#L90

carlescufi commented:
@ioannisg @stephanosio @mbolivar-nordic 

Running `check_compliance.py` and anything else in ci-tools is fully supported in downstream as @mbolivar-nordic pointed out, but we currently do not have a  clear deprecation policy for Kconfig options.
Lacking that, for now what I'd ask is that @stephanosio updates the release notes for 2.3 in `doc/releases` in this PR so that a note about this change is included. With that in I'd be happy to merge this.
